Define pseudo-code of header protection using packet number offset (#4436)
quicwg/base-drafts: draft-ietf-quic-tls.md

encoded on a single byte, or 2 bytes of frames for a 2-byte packet number
encoding.

The sampled ciphertext can be determined by the following pseudocode:

~~~

> sample_offset = packet_number_offset + 4

Reviewer: ```suggestion
# pn_offset is the start of the Packet Number field.
sample_offset = pn_offset + 4
```

We already use `pn_offset` in the header protection section above.

Author: Thank you for the suggestion. Applied as part of fd01f88.

encoded on a single byte, or 2 bytes of frames for a 2-byte packet number
encoding.

The sampled ciphertext can be determined by the following pseudocode:

~~~
# pn_offset is the start of the Packet Number field.
sample_offset = pn_offset + 4

sample = packet[sample_offset..sample_offset+sample_length]
~~~

where the packet number offset of a short header packet can be calculated as:

~~~
pn_offset = 1 + len(connection_id)
~~~

and the packet number offset of a long header packet can be calculated as:

~~~
pn_offset = 7 + len(destination_connection_id) +
                len(source_connection_id) +
                len(payload_length)
if packet_type == Initial:
    pn_offset += len(token_length) +
                 len(token)
~~~

For example, for a packet with a short header, an 8-byte connection ID, and
protected with AEAD_AES_128_GCM, the sample takes bytes 13 to 28 inclusive
(using zero-based indexing).

> Note that multiple QUIC packets might be included in the same UDP datagram and
> that each one is handled separately.

Reviewer 2: ```suggestion
Multiple QUIC packets might be included in the same UDP datagram. Each packet
is handled separately.
```
